### ISSUE TITLE
"date_shot" DB fields shouldn't be time zoned

### ISSUE THREAD

**Issue by ThibaultNocchi** [photoview/photoview#306]
**Describe the bug**
As seen [here](https://github.com/photoview/photoview/blob/master/api/scanner/exif/exif_parser_external.go#L58) and on various camera EXIF outputs, the creation dates aren't time zoned. This leads to having pictures times in the web interface converted to the browser's timezone and thus invalid times.

In Postgres DB at least, "date_shot" in "media" and "media_exif" are of type "timestampz", which is time zoned. This should be converted to a classic "timestamp" for it to display correct dates.

Edit: I also tried converting those two columns to "timestamp", and the web interface still displays a localized time so I guess it'll need to be changed :)

**To Reproduce**
1. Take a photo on your phone in a timezone other than UTC (it'll be like 4AM local time)
2. Import it, it'll be something like 4AM+0 in the DB
3. Check it in your localized browser, it'll display the picture being taken at 5AM+1 for instance

**Expected behavior**
Not timezone times, as EXIF shot time do not support that (and a lot of cameras don't)

**Screenshots**
#### Exiftool
![print_screen_2021-04-11-11-25-23](https://user-images.githubusercontent.com/1619359/114298859-a8b49900-9ab8-11eb-8242-bc31ccd1a352.png)
#### DB
![print_screen_2021-04-11-11-24-57](https://user-images.githubusercontent.com/1619359/114298860-a9e5c600-9ab8-11eb-946d-8f4b070a1a74.png)
#### Web UI
![print_screen_2021-04-11-11-24-41](https://user-images.githubusercontent.com/1619359/114298862-aa7e5c80-9ab8-11eb-8565-7f14dc187741.png)

**Your environment:**

- Postgres

**Comment by ggrzeczkowicz** [comment]
DateTimeOriginal is not timezoned, but the time zone could be contained in a lot of tag (TimeZoneOffset, GPSDateTime, TimeStamp....). Maybe it would be better to keep timestampz type, but to add a TimeZoneOffset column (as timestampz does not save timezoneoffset).

**Comment by ThibaultNocchi** [comment]
True, but if the app keeps the same EXIF tag as now, it doesn't have timezone so idk ^^

**Comment by ouafnico** [comment]
Any idea how to fix this timezone issue?

**Comment by coro1404** [comment]
tried this but to no avail:

```MariaDB [(none)]> SET time_zone = 'Europe/Berlin';
Query OK, 0 rows affected (0.000 sec)

MariaDB [(none)]> SELECT @@GLOBAL.time_zone, @@SESSION.time_zone;
+--------------------+---------------------+
| @@GLOBAL.time_zone | @@SESSION.time_zone |
+--------------------+---------------------+
| Europe/Berlin      | Europe/Berlin       |
+--------------------+---------------------+
1 row in set (0.000 sec)
```
Display times remain offset to exif data

**Comment by micha030201** [comment]
Note that this behavior is different between the internal and external parsers:

The internal parser does this: https://github.com/photoview/photoview/blob/124b6d34a93298bd84cb3a317855bd2f55af6a7d/api/scanner/exif/exif_parser_internal.go#L73-L78

The `DateTime` method, will, if the docs are to be believed, return the timestamp with the local timezone if the timezone information is not present: https://pkg.go.dev/github.com/xor-gate/goexif2@v1.1.0/exif#Exif.DateTime

The external parser, however, does this: https://github.com/photoview/photoview/blob/124b6d34a93298bd84cb3a317855bd2f55af6a7d/api/scanner/exif/exif_parser_external.go#L139-L159

The parsing is performed by `time.Parse`, which assumes UTC: https://pkg.go.dev/time#Parse

---

For my deployment, I've changed the external parser code to subtract the local offset when parsing the EXIF data to get the correct UTC timestamp for my photos:
```go
        //Get time of photo
        createDateKeys := []string{"CreationDate", "DateTimeOriginal", "CreateDate", "TrackCreateDate", "MediaCreateDate", "FileCreateDate", "ModifyDate", "TrackModifyDate", "MediaModifyDate", "FileModifyDate"}
        for _, createDateKey := range createDateKeys {
                date, err := fileInfo.GetString(createDateKey)
                if err == nil {
                        layout := "2006:01:02 15:04:05"
                        dateTime, err := time.Parse(layout, date)
                        if err == nil {
                                found_exif = true
                                _, offset := time.Now().Zone()
                                dateTimeUTC := dateTime.Add(-time.Duration(offset) * time.Second).UTC()
                                newExif.DateShot = &dateTimeUTC
                        } else {
                                layoutWithOffset := "2006:01:02 15:04:05-07:00"
                                dateTime, err = time.Parse(layoutWithOffset, date)
                                if err == nil {
                                        found_exif = true
                                        newExif.DateShot = &dateTime
                                }
                        }
                        break
                }
        }
```

This works for me because my server is set to the same timezone as my camera. I could perhaps do it more cleanly by actually writing the timezone instead of subtracting the offset, but it works well enough.

This may not work for everyone, because it's not good to assume in general that the camera and the server are set to the same timezone.

This brings up the more important question about the behavior though -- I personally would rather all the timestamps be displayed in the timezone *they were taken in* -- if I go somewhere with a different timezone and take pictures there, it serves me absolutely no purpose to see in the web interface what time it was where I currently am when that picture was taken, I would rather it showed the time from the camera as-is, maybe with the timezone information displayed alongside it if present. But that's just, like, my opinion.

Changing the behavior to that would require touching the frontend code, and probably adding another field to the database to differentiate between the timezone being set to UTC and not being present. This seemed substantially more difficult, so I didn't do it.

**Comment by kkovaletp** [comment]
@micha030201, thanks for bringing this case up.
I did some research on the code ( #1253 ) and found that most cameras don't store the timezone in the EXIF data, but put a date-time value in the UTC timezone. Our code already works fine with such media files.
However, there are some edge cases where cameras might store the date-time with the local timezone. And for this particular case, we have an inconsistency in the code.

So far, these are theoretical assumptions and extrapolations: I tried to create a test image with the modified EXIF data, containing the timezone, but failed to do that in a way, Photoview expects it to get from the file.

Can you please attach a zipped example of a picture with the EXIF dates, stored with a timezone? I'll test our code with your image to prove that we have a bug in the code.

**Comment by kkovaletp** [comment]
There is no response for a month, also I wasn't able to create a test image to reproduce the issue, and I'm not sure that it is even possible...
So, I'm closing this bug as completed, assuming that at the time it was created, those fields were handled incorrectly, and later the behavior was fixed.
